File: CSharp/AoC2025/Days/Day4.cs
namespace AoC2025.Days;

public static class Day4
{
    public static int PartOne()
    {
        return TraverseGrid(SetUpGrid());
    }

    public static int PartTwo()
    {
        char[][] grid = SetUpGrid();
        int result = 0;
        while (true)
        {
            int total = TraverseGrid(grid, true);
            if (total == 0)
                break;
            result += total;
        }
        return result;
    }


    // Helper methods

    private static int TraverseGrid(char[][] grid, bool removeRolls = false)
    {
        int height = grid.Length;
        int width = grid[0].Length;
        int result = 0;

        for (int i = 1; i < height - 1; i++)
        {
            for (int j = 1; j < width - 1; j++)
            {
                if (grid[i][j] == '.')
                    continue;
                if (CountNeighbours(i, j, grid) < 4)
                {
                    result++;

                    if (removeRolls)
                        grid[i][j] = '.';
                }
            }
        }
        return result;
    }

    private static int CountNeighbours(int row, int col, char[][] grid)
    {
        int[] deltas = [-1, 0, 1];
        int total = grid[row][col] == '@' ? -1 : 0;
        foreach (int i in deltas)
        {
            foreach (int j in deltas)
            {
                if (grid[row + i][col + j] == '@')
                {
                    total++;
                }
            }
        }
        return total;
    }
    
    // Data methods

    private static string[] GetData()
    {
        return File.ReadAllLines("../../Data/2025/day4.txt");
    }

    private static char[][] SetUpGrid()
    {
        string[] data = GetData();
        string emptyLine = new string('.', data[0].Length + 2);
        return new[] {emptyLine.ToCharArray()}
            .Concat(
                data
                .Select(line => ('.' + line + '.').ToCharArray())
                .ToArray()
            )
            .Concat(
                new[] {emptyLine.ToCharArray()}
            )
            .ToArray();
    }
}

File: JavaScript/AoC2025/days/day4.js
import fs from "fs"
import path from "path"

export default class Day4 {
    constructor() {
        const filePath = path.resolve("../../Data/2025/day4.txt");
        const data = fs.readFileSync(filePath, "utf-8");
        this.lines = data.split('\n').map(line => line.trimEnd());
        this.height = this.lines.length;
        this.width = this.lines[0].length;
        this.grid = this.padGrid();
    }

    printResults() {
        console.log("Day 4:");
        console.log(`Part 1: ${this.part1()}`);
        console.log(`Part 2: ${this.part2()}`);
    }

    part1() {
        let result = 0;
        for (let i = 1; i < this.height + 1; i++) {
            for (let j = 1; j < this.width + 1; j++) {
                if (this.grid[i][j] === '@' && this.countNeighbours(i, j) < 4) {
                    result++;
                }
            }
        }
        return result;
    }

    part2() {
        let result = 0;
        while (true) {
            let total = 0;
            for (let i = 1; i < this.height + 1; i++) {
                for (let j = 1; j < this.width + 1; j++) {
                    if (this.grid[i][j] === '@' && this.countNeighbours(i, j) < 4) {
                        total++;
                        this.grid[i][j] = '.';
                    }
                }
            }
            result += total;
            if (total === 0)
                break;
        }
        return result;
    }

    padGrid() {
        const array = []
        const emptyLine = Array(this.width + 2).fill('.');
        array.push([...emptyLine]);
        for (const line of this.lines) {
            array.push(("." + line + ".").split(""));
        }
        array.push([...emptyLine]);
        return array;
    }

    countNeighbours(row, col) {
        let total = 0;
        for (let i = -1; i <= 1; i++) {
            for (let j = -1; j <= 1; j++) {
                if (this.grid[row + i][col + j] == '@')
                    total++;
            }
        }
        return total - 1; // Ignore the @ in row + 0, col + 0
    }
}